cogent/src/Cogent/DataLayout/Core.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE DeriveFunctor #-}
+module Cogent.DataLayout.Core where
+  
+import Data.Map (Map)
+
+import Text.Parsec.Pos (SourcePos)
+
+import Cogent.Common.Syntax (TagName, FieldName)
+
+
+{- CORE DATALAYOUT TYPES -}
+
+-- A range of bit indices into a data type.
+--
+-- Should satisfy `bitSizeBR >= 1` and `bitOffset >= 0`.
+-- Represents the set {bitOffset, bitOffset + 1, ..., bitOffset + bitSize - 1}
+data BitRange
+  = BitRange { bitSizeBR :: Integer, bitOffsetBR :: Integer }
+  deriving (Eq, Show, Ord)
+  
+-- A range of bit indices into a data type.
+--
+-- Should satisfy `bitSizeABR >= 1`, `bitOffsetABR >= 0` and `wordOffsetABR >= 0`.
+-- Should be aligned in the sense that `bitSizeABR + bitOffsetABR <= wordSizeBits`.
+-- Represents the set
+-- { wordOffset * wordSizeBits + bitOffset
+-- , wordOffset * wordSizeBits + bitOffset + 1
+-- , ...
+-- , wordOffset * wordSizeBits + bitOffset + bitSize - 1
+-- }
+data AlignedBitRange
+  = AlignedBitRange { bitSizeABR :: Integer, bitOffsetABR :: Integer, wordOffsetABR :: Integer }
+  deriving (Eq, Show, Ord)
+
+-- A specification of how to get/set the bits for part of a primitive type
+--
+-- Size 0 surface language blocks go to core language UnitLayout rather than PrimLayout.
+-- After splitting, the BitRange of DataBlocks cannot contain a multiple of the wordSize.
+-- This ensures that each block is fully contained within an aligned word.
+-- All heap allocated structures must be a multiple of a word in size. Malloc guarantees this by default???
+data Block
```

Comment:
I think my original intention was for `alignLayout` to map to `[Block]` rather than `[AlignedBitRange]`. If the size of the original BitRange is smaller than the size of the type that the PrimLayout is matched with, will need to add padding to bring it up to the same size. Not sure if we need to express this in the data structure though?

In the future, if we expose into the surface language the ability to decompose the layout of primitive types as a list of bit ranges in this way, then it would also make sense to expose the ability to add padding in the surface language as well.